util/reflect_test.go

```diff
@@ -856,6 +911,103 @@ func TestForEachField(t *testing.T) {
 	}
 }
 
+func TestForEachDataField(t *testing.T) {
+	tests := []struct {
+		desc         string
+		parentStruct interface{}
+		in           interface{}
+		out          interface{}
+		iterFunc     FieldIteratorFunc
+		wantOut      string
+		wantErr      string
+	}{
+		{
+			desc:         "nil",
+			parentStruct: nil,
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut:      ``,
+		},
+		{
+			desc:         "nil fields",
+			parentStruct: &PointerOnlyBasicStruct{},
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut:      ``,
+		},
+		{
+			desc:         "invalid path specification for field",
+			parentStruct: &PathErrorStruct{Field: String("value")},
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut:      ``,
+			wantErr:      "field Field did not specify a path",
+		},
+		{
+			desc:         "struct without schema",
+			parentStruct: &basicStruct1,
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut:      `int32 : 42, string : "forty two", int32ptr : 4242, stringptr : "forty two ptr", `,
+		},
+		{
+			desc:         "struct of struct with no schema",
+			parentStruct: &StructOfStructs{BasicStructField: basicStruct1, BasicStructPtrField: &basicStruct2},
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut: `int32 : 42, string : "forty two", int32ptr : 4242, stringptr : "forty two ptr", ` +
+				`int32 : 43, string : "forty three", int32ptr : 4343, stringptr : "forty three ptr", `,
+		},
+		{
+			desc:         "struct of slice of structs with no schema",
+			parentStruct: &StructOfSliceOfStructs{BasicStructSliceField: []BasicStruct{basicStruct1}, BasicStructPtrSliceField: []*BasicStruct{&basicStruct2}},
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut: `int32 : 42, string : "forty two", int32ptr : 4242, stringptr : "forty two ptr", ` +
+				`int32 : 43, string : "forty three", int32ptr : 4343, stringptr : "forty three ptr", `,
+		},
+		{
+			desc:         "struct with string leaf-list",
+			parentStruct: &BasicSliceStruct{StringSlice: []string{"one", "two"}},
+			in:           nil,
+			iterFunc:     printSchemaAnnotationFieldsIterFunc,
+			wantOut:      ``,
```

Comment:
Why do we want an empty output here?

If that is intentional, can we update the test case description to make it more obvious?

Reply:
This is an error -- and a bug -- thanks for catching it. The error is just in the test iteration function since `IsValueScalar` check was wrong, the overall implementation was calling the right thing.